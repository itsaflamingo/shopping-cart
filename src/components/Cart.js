import React, {useState, useEffect} from 'react';
import { useNavigate } from 'react-router-dom';
import Items from './Items';
import {itemsMade} from './shop-items';


function AddToCart(props) {

  const [quantity, setQuantity] = useState(0);

  const updateQuantity = () => {
    const increaseQuantity = (e, num) => {
      setQuantity(num+1)
      console.log(quantity)
      updateObj(e)
    }
    const decreaseQuantity = (num) => setQuantity(num-1);
    const updateObj = (e) => {
      if(e === undefined) return;
      const name = e.target.name;
      console.log(e.target.name);
      itemsMade.forEach(item => {
        if(item.name === name) {
          item.quantity = quantity;
        }
        else {
          return;
        }
        console.log(item)
      })
    }

    return {
      increaseQuantity,
      decreaseQuantity,
      updateObj
    }
  }

  return (
    <div id='cart' role='cart' aria-describedby="modal">
      <div>
        <h3>Shopping Cart</h3>
        <button>Close</button>
      </div>
      {props.items.map((item) => {
        return(
          <div className='cart-item' key={item.id}>
            <div className='picture' id='cart-picture'></div>
            <h3 className='item-name'>{item.name}</h3>
            <h4 className='price'>{item.price}</h4>
            <div id='item-counter'>{item.quantity}</div>
            <div id='plus-minus-item'>
              <button id='add' name={item.name} onClick={(e) => updateQuantity().increaseQuantity(e, quantity)}>+</button>
              <button id='remove' onClick={() => updateQuantity().decreaseQuantity(quantity)}>-</button>
            </div>
          </div>
        )
      })}
    </div>
  );
}

export default AddToCart;
