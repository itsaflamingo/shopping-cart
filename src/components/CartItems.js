import React from 'react';
import Counter from './Counter'


export default function CartItems(props) {

    const {item, newTotal} = props;

  return (
          <div className='cart-item'>
            <div className='picture' id='cart-picture'
            style={{
              backgroundImage: `url(${item.image})`,
              backgroundSize: 'contain',
              backgroundRepeat: 'no-repeat'
            }}></div>
            <h4 className='item-name'>{item.title}</h4>
            <h4 className='price'>${item.price}</h4>
            <Counter item={item} newTotal={newTotal} />
          </div>
  );
}
