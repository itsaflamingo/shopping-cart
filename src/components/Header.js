import React, {useEffect, useState} from 'react';
import shoppingCart from '../images/shopping-cart.png'
import Cart from './Cart'
import home from '../images/home.png'
import search from '../images/search.png'
import Search  from './Search';

export default function Header(props) {

    const [closeCart, setCloseCart] = useState(true);
    const [openSearch, setOpenSearch] = useState(false);
    const {total, newTotal, cart, navHome} = props;

  return (
      <div id='shop-header'>
        <h2>Shop</h2>
        <div id='buttons'>
        <div id='search' style={{
                backgroundImage: `url(${search})`,
                backgroundSize: 'contain',
                backgroundRepeat: 'no-repeat'
              }}></div>
        <div id='to-home' onClick={navHome} style={{
                backgroundImage: `url(${home})`,
                backgroundSize: 'contain',
                backgroundRepeat: 'no-repeat'
              }}></div>
        <div id='cart-icon' style={{
                backgroundImage: `url(${shoppingCart})`,
                backgroundSize: 'contain',
                backgroundRepeat: 'no-repeat'
              }} onClick={() => setCloseCart(!closeCart)}>
                <div id='cart-quantity'>{cart.length}</div>
            </div>
        </div>
        {!closeCart && <Cart items={cart} setCloseCart={setCloseCart} total={total} newTotal={newTotal} />}
        {openSearch && <Search />}
      </div>
  );
}