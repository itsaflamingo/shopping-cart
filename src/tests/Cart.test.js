import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";  // optional
import {BrowserRouter as Router} from 'react-router-dom';
import { act } from "react-dom/test-utils";
import userEvent from "@testing-library/user-event";
import Cart from '../components/Cart'
import CartItems from '../components/CartItems'
import Header from '../components/Header'

// renders html correctly - snapshot
describe('Cart component', () => {
    const setCloseCart = jest.fn()
    const navHome = jest.fn()
    const items = [
        {
            image: 'url1',
            title: 'name1',
            price: '10',
            id: 1   },
        {
            image: 'url2',
            title: 'name2',
            price: '10',
            id: 2   },
        {
            image: 'url3',
            title: 'name3',
            price: '10',
            id: 3   }
    ]
    const item = {
        image: 'url1',
        title: 'name1',
        price: '10',
        id: 1   }
    const newTotal = jest.fn();
    const total = 0;
    it('Renders cart on page', () => {
        const {container} = render(
            <Router>
                <Cart items={items} setCloseCart={setCloseCart} newTotal={newTotal} total={total} />
            </Router>) 
            
            expect(container).toMatchSnapshot();
    })
    it('Cart total starts at 0', () => {
        render(
            <Router>
                <Cart items={items} setCloseCart={setCloseCart} newTotal={newTotal} total={total} />
                <CartItems item={item} newTotal={newTotal} />
            </Router>) 
        
        const input = screen.getByLabelText('total');
        expect(input.innerHTML).toBe('Total: $0');
    })
    it('Cart total increases by correct amount when number of cart items increases', () => {
        let thisTotal = 0;
        items.forEach(item => {
            thisTotal = thisTotal + parseInt(item.price);
        })
        render(
            <Router>
                <Header navHome={navHome} cart={items} newTotal={newTotal} total={thisTotal} />
                <CartItems item={item} newTotal={newTotal} />
            </Router>) 

        const button = screen.getByTestId('cart-icon');
        act(() => {
            userEvent.click(button);
        })
        
        const input = screen.getByLabelText('total');
        expect(input.innerHTML).toBe(`Total: $30`);
    })
    test.todo('Cart total decreases by correct amount when - is clicked')
    test.todo('Cart total increases by correct amount when item is added to cart')
    test.todo('when item quantity is changed, total price is changed')
    test.todo('when new item is added, total price is changed')
    test.todo('total price calculation is correct')
})
