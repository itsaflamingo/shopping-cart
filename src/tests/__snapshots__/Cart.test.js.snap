// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component Renders cart on page 1`] = `
<div>
  <div
    aria-describedby="modal"
    id="cart"
    role="cart"
  >
    <div
      id="cart-header"
    >
      <h3>
        Shopping Cart
      </h3>
    </div>
    <div>
      <div
        class="cart-item"
      >
        <div
          class="cart-picture"
          style="background-image: url(url1); background-size: contain; background-repeat: no-repeat;"
        />
        <div
          class="cart-item-info"
        >
          <h4
            class="item-name"
          >
            name1
          </h4>
          <h4
            class="price"
          >
            $
            10
          </h4>
          <div
            class="counter"
          >
            <input
              class="item-counter"
              type="input"
              value=""
            />
            <div
              class="item-quantity"
            >
              <button
                class="add"
              >
                +
              </button>
              <button
                class="remove"
              >
                -
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="cart-item"
      >
        <div
          class="cart-picture"
          style="background-image: url(url2); background-size: contain; background-repeat: no-repeat;"
        />
        <div
          class="cart-item-info"
        >
          <h4
            class="item-name"
          >
            name2
          </h4>
          <h4
            class="price"
          >
            $
            10
          </h4>
          <div
            class="counter"
          >
            <input
              class="item-counter"
              type="input"
              value=""
            />
            <div
              class="item-quantity"
            >
              <button
                class="add"
              >
                +
              </button>
              <button
                class="remove"
              >
                -
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="cart-item"
      >
        <div
          class="cart-picture"
          style="background-image: url(url3); background-size: contain; background-repeat: no-repeat;"
        />
        <div
          class="cart-item-info"
        >
          <h4
            class="item-name"
          >
            name3
          </h4>
          <h4
            class="price"
          >
            $
            10
          </h4>
          <div
            class="counter"
          >
            <input
              class="item-counter"
              type="input"
              value=""
            />
            <div
              class="item-quantity"
            >
              <button
                class="add"
              >
                +
              </button>
              <button
                class="remove"
              >
                -
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-label="total"
      id="total"
    >
      Total: $
      0
    </div>
    <button
      id="checkout"
    >
      Checkout
    </button>
  </div>
</div>
`;
